include/zephyr/sys/sem.h

```diff
@@ -138,6 +139,82 @@ int sys_sem_take(struct sys_sem *sem, k_timeout_t timeout);
  */
 unsigned int sys_sem_count_get(struct sys_sem *sem);
 
+/**
+ * @cond INTERNAL_HIDDEN
+ */
+
+#if defined(__GNUC__)
+static ALWAYS_INLINE void z_sys_sem_lock_onexit(__maybe_unused int *rc)
+{
+	__ASSERT(*rc == 1, "SYS_SEM_LOCK exited with goto, break or return, "
+			   "use SYS_SEM_LOCK_BREAK instead.");
+}
+#define SYS_SEM_LOCK_ONEXIT __attribute__((__cleanup__(z_sys_sem_lock_onexit)))
```

Comment:
Seems like if we're going to repeat this paradigm we should think about up-levelling it so that an abstracted "lock" can provide the preprocessor tools to define these loop/block macros with  ~one-line defintions or whatever.  But that's for a different PR.